### PR TITLE
Multiple runners, configurable runners and extra options

### PR DESCRIPTION
**Pull request by paulrbr-fl**
This PR is quite a big change but was necessary for us at [`Fretlink`](https://github.com/fretlink/).

The 3 main changes are:
- allows to register multiple runners on the same instance
- allows to **remove** an existing runner with the `gitlab_runner_runners[i].state` value (`present` or `absent`)
- allows to change configuration for a runner (after registration) fixing #17 
- allow extra TOML options to be configured 
  Adds a `extra_configs` key to a runner's variables to
allow custom TOML sections to be changed/added. 
  The operation is **non-destructive**. Meaning that any existing
configuration option will **not** be removed if they already exist in
the `/etc/gitlab-runner/config.toml` file.

### CODE REVIEW

**Comment by riemers** [comment]
This is a nice and solid addition 👍 

Since this will break current installations if they auto download this version it might be wise to make a note or similar. I can change the title of the repo to include it. Will bump it with a new major version too. Is that sufficient?

**Comment by paulrbr-fl** [comment]
> This is a nice and solid addition +1
> 
> Since this will break current installations if they auto download this version it might be wise to make a note or similar. I can change the title of the repo to include it. Will bump it with a new major version too. Is that sufficient?

Hi @riemers. Agreed it sounds like a wise thing to do. Adding a notice in the README too when you major version bump the role :).

I have done many tests on my side, so hopefully I should not break any existing features. If you can do some tests on your side it would be perfect.

Thanks for the review!

**Comment by paulrbr-fl** [comment]
Hi @rgevaert, thanks for your comment on https://github.com/riemers/ansible-gitlab-runner/issues/33#issuecomment-456824198. I removed the mandatory `state` field on a runner config and made it `'present'` by default.

For your second issue I am surprised you had to add  `gitlab_runner_docker_privileged` as it's not used anywhere in the role! Are you sure you took the latest version of code of this PR?

**Comment by rgevaert** [comment]
Thanks, I didn't find this thread. Now I switched to your multiple-runners branch. Let me try again. The no_log: true option is also slowing down debugging  and updating the code ;)

**Comment by rgevaert** [comment]
```
task path: /Users/rgevaert/Vagrant/ubuntu18/provisioning/roles/riemers.gitlab-runner/tasks/register-runner.yml:2
fatal: [default]: FAILED! => {
    "msg": "The task includes an option with an undefined variable. The error was: 'dict object' has no attribute 'run_untagged'\n\nThe error appears to have been in '/Users/rgevaert/Vagrant/ubuntu18/provisioning/roles/riemers.gitlab-runner/tasks/register-runner.yml': line 2, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n---\n- name: Register runner to GitLab\n  ^ here\n"
}
```
setting it per runner     run_untagged works, but it is not getting the defaults.

**Comment by paulrbr-fl** [comment]
> ```
> task path: /Users/rgevaert/Vagrant/ubuntu18/provisioning/roles/riemers.gitlab-runner/tasks/register-runner.yml:2
> fatal: [default]: FAILED! => {
>     "msg": "The task includes an option with an undefined variable. The error was: 'dict object' has no attribute 'run_untagged'\n\nThe error appears to have been in '/Users/rgevaert/Vagrant/ubuntu18/provisioning/roles/riemers.gitlab-runner/tasks/register-runner.yml': line 2, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n---\n- name: Register runner to GitLab\n  ^ here\n"
> }
> ```
> 
> setting it per runner run_untagged works, but it is not getting the defaults.

Indeed, Thanks for checking! I have fixed my branch by using a default value if it's not defined.

**Comment by riemers** [comment]
Thanks for working on this one guys.. if you both give the greenlight i will update the role on galaxy.

**Comment by paulrbr-fl** [comment]
@rgevaert thanks for all the feedbacks. I have corrected the errors you raised. 

I also made sure to run the playbook in the automatic tests. (I cannot test a real gitlab runner registration though unfortunately).

The branch seems solid now :)

**Comment by riemers** [comment]
Are we good to go with merging this?

**Comment by paulrbr-fl** [comment]
@riemers we use it in my company regularly. I looks good. @rgevaert do you have any more feedback? 🙂

**Comment by riemers** [comment]
Released it as a major version, updated the main text of the repo to include it too. Did some trials too and looked fine for me. If you have the chance, you can ofcourse check it via galaxy.
